src/ORM/Query.php

```diff
@@ -1399,4 +1860,66 @@ public function isAutoFieldsEnabled(): ?bool
     {
         return $this->_autoFields;
     }
+
+    /**
+     * Will iterate over every specified part. Traversing functions can aggregate
+     * results using variables in the closure or instance variables. This function
+     * is commonly used as a way for traversing all query parts that
+     * are going to be used for constructing a query.
+     *
+     * The callback will receive 2 parameters, the first one is the value of the query
+     * part that is being iterated and the second the name of such part.
+     *
+     * ### Example
+     * ```
+     * $query->select(['title'])->from('articles')->traverse(function ($value, $clause) {
+     *     if ($clause === 'select') {
+     *         var_dump($value);
+     *     }
+     * });
+     * ```
+     *
+     * @param \Closure $callback Callback to be executed for each part
+     * @return $this
+     */
+    public function traverse(Closure $callback)
+    {
+        $this->dbQuery()->traverse($callback);
+
+        return $this;
+    }
+
+    /**
+     * Magic method for proxing calls to internal database query instance.
+     *
+     * @param string $name Method name.
+     * @param array $arguments Arguments.
+     * @return mixed
+     */
+    public function __call(string $name, array $arguments): mixed
```

Comment:
Are you planning to add `@method` or `@mixin` hints to account for the "missing" methods that are being proxied here?

Reply:
Yes, will have to add a ton of `@method` annotations to the class :(

Reply:
\**Cries in Spanish*\*

Reply:
Hours later I'm still sad... Losing the API documentation and getting stuck in `@method` annotations when navigating through declarations is always such a pain ;(

I had doubts about `@mixin` support, but I saw that the core already uses it (though only once with `\Cake\ORM\Association`), and from a quick look it appears that IDE support is actually quite OK nowadays, so maybe we could get away with that instead?

```php
 * @mixin \Cake\Database\Query\SelectQuery
 * @mixin \Cake\Database\Query\InsertQuery
 * @mixin \Cake\Database\Query\UpdateQuery
 * @mixin \Cake\Database\Query\DeleteQuery
```

Reply:
The key issue with @method annotations is the format allows for basically just a single summary line.

Reply:
@ndm2 Can't use @mixin as all the fluent builder methods return `$this` so the static analyzers think the return value is of the db query type and not ORM\Query.

Reply:
Are you sure? That would be really stupid of them. I gotta give it a try, I remember the problems with `self`, but `$this` should clearly be the current object's type.

Reply:
Yup I tried and got the behavior I mentioned with phpstan. But please try yourself too.

Reply:
OK this sucks big time, how is this not a bug in PHPStan? Clearly this is not how `@mixin` actually works in the real world ;(

Anyhow, how about we handle this with an extension?

Reply:
> Anyhow, how about we handle this with an extension?

Sure if someone wants to take a crack at it.

Reply:
I'll give it a try if we're generally cool with that route. I saw that there already are extensions, and I guess developers that run into problems are encouraged to use them for their own stan runs?

Reply:
Would be nice if this PHPStan rule could be added to `CakeDC/cakephp-phpstan` since it already provides CakePHP specific rules and is fairly established in our community. But this is of course up to you.

Reply:
@LordSimal If we go with an extension, and they create a 5.x variant, sure, why not, they already copied over the other extensions from the core.

@ADmad This should be pretty close to how the proxy works: https://github.com/ndm2/cakephp/pull/26/commits/4119a212c34a93fac84da622634ecd882615cb94